Implement parameter validation testing across all server routes

omarraf opened the pull request:
# Description

This PR implements comprehensive parameter validation testing across all server routes as requested in #1497. The testing framework systematically validates that each route properly handles invalid parameters, preventing malformed requests from reaching application logic.

## Key changes
  ### 1.) Reusable Testing Framework

  Created src/server/test/util/validationHelpers.js with helper functions for:
  - String validation (length constraints, enum values)
  - Integer/boolean validation (type checking, bounds)
  - Token security testing (DoS prevention)
  - Parameter injection prevention (extra field rejection)
  - Malicious input testing (SQL injection, XSS, path traversal)

  ### 2.) Centralized Validation Constants

  Added src/server/util/validationConstants.js to eliminate hardcoded validation limits and ensure
  consistency across routes

  ### 3.) Comprehensive Test Coverage

  Added 20+ parameter validation test files covering:
  - Core Resources: units, meters, groups, maps
  - Data & Readings: readings, compare, baseline, CSV import/export
  - Conversions: unit conversions and bulk operations
  - Auth: login, 2FA, user management, verification
  - System: preferences, logs, version, CIKS, Obvius integration


## Related
This work builds upon  #1512

Fixes #1497 

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

omarraf committed:
add testing for conversions.js

omarraf committed:
fix failing tests for conversion tests

omarraf committed:
authenticator.js param validation and tests

omarraf committed:
login param validation and tests

omarraf committed:
users.js param validation and tests

omarraf committed:
meters testing and validation

omarraf committed:
groups param validation and testing

omarraf committed:
compareReadings param validation and test file

omarraf committed:
maps.js param validation and testing

omarraf committed:
update conversions to use framework

omarraf committed:
preferences.js param validation and testing

omarraf committed:
csv.js test file

omarraf committed:
ciks.js test file

omarraf committed:
version.js test

omarraf committed:
improve paramter validation

omarraf committed:
response.js test file

omarraf committed:
create readings.js param test file

omarraf committed:
verification param test file

omarraf committed:
obvius.js param test file

omarraf committed:
conversionArray.js param validation and test file

omarraf committed:
baseline test file

omarraf committed:
temporary validation helper file

omarraf commented:
Hi @huss, I created this draft PR to get some feedback on the approach before continuing further.

  Temporary accommodations made:
  - Some tests are commented out due to failing 
  - A few test expectations are broadened to handle auth-first middleware flows
  - Some security limits are conservative estimates vs. actual database constraints

  Few questions I had:
  1. Overall approach - does this testing methodology look good?
  2. Security limits - are the hardcoded values reasonable? (1000 char strings, etc.)
  3. Any other immediate concerns?

I just wanted to make sure I'm on the right track before continuing, thanks!

huss commented on src/server/routes/authenticator.js:
```diff
@@ -19,7 +19,8 @@ const escapeHtml = require('escape-html');
 	const validParams = {
-		type: 'string'
+		type: 'string',
+		maxLength: 2000
```
I'm not sure how big this is but I assume 2000 is safe.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -59,7 +61,8 @@ function credentialsRequestValidationMiddleware(req, res, next) {
 				type: 'string',
-				minLength: 8
+				minLength: 8,
+				maxLength: 1000
```
This is not designed to test the normal application imposed limit but to avoid misuse that could cause issues. Given this, I have two thoughts:

- I think the value should be defined in a utility file. I don't see an obvious place for it in src/server/util so a new file might be good. It could have a general name so other constants can be added later. This way all uses can easily be changed.
- I'm not certain what the limit should be. I'm thinking that across routes a string limit in the 256-1024 would be good. If it is a global constant then it is easy to change. Having seen it is used for notes then may the larger end is better.

omarraf replied:
> This way all uses can easily be changed.

I agree, that does sound like a better approach. I'll work on improving this.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -50,6 +51,7 @@ authMiddleware = (req, res, next) => {
 function credentialsRequestValidationMiddleware(req, res, next) {
 	const validParams = {
 		type: 'object',
+		maxProperties: 2,
```
@maigbao found that ``additionalProperties: false`` was a stronger and easier restriction when only the required params are allowed. It replaces the ``maxProperties``. If you concur then it can be changed. Later I saw you did that for refresh.

huss commented on src/server/routes/baseline.js:
```diff
@@ -17,7 +19,47 @@ router.get('/', async (req, res) => {
 	}
 });
-router.post('/new', async (req, res) => {
+router.post('/new', adminAuthenticator('create baselines'), async (req, res) => {
```
FYI. This is an old file that is currently being redone. It is fine to put in your changes but the other team will, likely, also make changes.

omarraf replied:
I see, do you think it'll be better to leave it alone for now to avoid confusion?

huss replied:
As long as you use the correct authenticator, I think it is fine to update. When your work is merged with their work (I assume after you are done) then any conflict can be resolved.

huss commented on src/server/routes/baseline.js:
```diff
@@ -17,7 +19,47 @@ router.get('/', async (req, res) => {
+		type: 'object',
+		additionalProperties: false,
+		required: ['meterID', 'applyStart', 'applyEnd', 'calcStart', 'calcEnd'],
+		properties: {
```
Should this have a maxProperties?

huss commented on src/server/routes/baseline.js:
```diff
@@ -17,7 +19,47 @@ router.get('/', async (req, res) => {
+			meterID: {
+				type: 'integer',
+				minimum: 1,
+				maximum: 2147483647
```
I'm unsure if this is needed or helps. I tried to find the maximum imposed by JSON Schema but did not locate. If someone sent a bad value then the DB would reject it. I don't think having a big value changes that. It is also unclear Postgres limits the key values it generates so I don't think OED can set a lower limit on the maximum. Thus, I'm inclined not to have a maximum unless others have a different opinion.

The minimum, on the other hand, stops bad usage and is fine.

huss commented on src/server/routes/baseline.js:
```diff
@@ -17,7 +19,47 @@ router.get('/', async (req, res) => {
+			},
+			applyStart: {
+				type: 'string',
+				maxLength: 100
```
I'm inclined to use the same global max on all strings unless there is a specific reason not to (for example, the JSON token length that you did). I think it provides the same level of protection and is less likely to fail for valid uses.

Now, if there was a way to systematically verify that these are valid date/time stamps then that would be valuable. OED uses them in a number of places. I think moment can tell you but I'm uncertain that the JSON schema testing can. If not then it would need to happen after that testing with an OED defined function. OED already has this type of testing outside the JSON schema proposed in @maigbao pull request and I'm okay with continuing that if you wanted to do this (or leave a TODO for the future).

huss commented on src/server/routes/compareReadings.js:
```diff
@@ -50,16 +52,20 @@ function validateQueryParams(queryParams) {
 			},
 			graphicUnitId: {
 				type: 'string',
+				maxLength: 20,
```
I might suggest a separate global const on the string length to mirror string as this should be used in several routes. This should be big enough for a 64-bit integer so the value seems fine.

huss commented on src/server/routes/conversionArray.js:
```diff
@@ -6,21 +6,51 @@ const express = require('express');
+		return;
+	}
+	
+	try {
```
I like that you are dealing with any potential error. However, all uses need to be checked if it will be okay with 500. I think what really should happen is the standard error function should be used as in other places. Unless this is needed for your testing, I would make this a TODO to deal with the error condition and leave it for future work. You can do it as a different PR if you want.

huss replied:
If I'm looking correctly at the latest code, the new param testing and try was removed. I propose to add back:

```js
	const validParams = {
		type: 'object',
		additionalProperties: false,
		properties: {
			redoCik: {
				type: 'boolean'
			},
			refreshReadingViews: {
				type: 'boolean'
			}
		}
	};
	
	if (!validate(req.body, validParams).valid) {
		res.sendStatus(400);
		return;
	}
```
Note the added import for validate is not needed now but will be with this code put back.

For the other test of
```
	// At least one operation must be requested
	if (!req.body.redoCik && !req.body.refreshReadingViews) {
		res.sendStatus(400);
		return;
	}
````
I think that can be left out. It probably is okay but OED might send a request to do nothing in special cases.

Could you also add a TODO comment to deal with failures with a try/catch?

huss commented on src/server/routes/conversions.js:
```diff
@@ -33,33 +34,38 @@ router.get('/', async (req, res) => {
  * Route for POST, edit conversion.
  */
-router.post('/edit', async (req, res) => {
+router.post('/edit', adminAuthenticator('edit conversions'), async (req, res) => {
```
I'm wondering why the recent changes to many routes for the authentication is not showing up in your pull request. This was fixed (and a number of others) but used a different function that OED standardized on. I'm sorry but these changes to route middleware need to go and the new changes incorporated unless you find an issue with the new code. Given this, I'm not going to comment much more on the route changes until this is resolved.

omarraf replied:
That's my bad, I'll rebase against main to pick up the authentication changes and update the PR.

huss commented on src/server/routes/conversions.js:
```diff
@@ -33,33 +34,38 @@ router.get('/', async (req, res) => {
+router.post('/edit', adminAuthenticator('edit conversions'), async (req, res) => {
 	const validConversion = {
 		type: 'object',
+		maxProperties: 6, // Prevent parameter injection attacks
```
First, I'm unclear on why the comment was added here and not elsewhere. I'm not advocating for putting it in every place but maybe none would be okay? Second, OED prefers comments on the line above the code if one is going to stay.

omarraf replied:
No particular reason, I'll remove it to keep things consistent. I'll also move any remaining comments above the code

huss commented on src/server/routes/obvius.js:
```diff
@@ -125,11 +131,22 @@ function verifyObviusUser(req, res, next) {
 	}
+	
+	// Basic parameter validation for security
+	if (typeof password !== 'string' || password.length > 1000) {
```
The original logic had if/else if/else. While it should not matter, I prefer that. In that usage each case would be else if until the else for the last step (not in if/else now). I think it is clearer that it should only happen if all the tests pass. I also think all the returns are no longer needed since they are each unique cases (I know the old code did have the returns). Thoughts?

omarraf replied:
Yeah that sounds like a good idea, it is more clear and readable as an if/else. I'll update this

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+
+const { expect } = require('chai');
+const { chai, mocha, app, testDB } = require('../common');
+const { validateString, validateToken, testInvalidField } = require('../util/validationHelpers');
```
validateToken seems unused.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+                basePayload: baseCredentials,
+                required: true,
+                minLength: 5,
+                maxLength: 254
```
If there are global max for the param testing to set limits then they should be used here. This applies to all similar uses.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+                endpoint: LOGIN_ENDPOINT,
+                basePayload: baseCredentials,
+                required: true,
+                minLength: 5,
```
I think this is out-of-date and is now 3 in development.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+            const res = await chai.request(app)
+                .post(LOGIN_ENDPOINT)
+                .send(payloadWithExtra);
+            expect(res).to.have.status(400);
```
OED defines code values in src/server/util/readingsUtils.js. It might be good to integrate them in your tests even though OED has not systematically done that.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+            expect(res).to.have.status(400);
+        });
+
+        mocha.it('should handle malicious username inputs', async () => {
```
I like the idea of these tests. I'm wondering if the test should try to pass a command in the username, password, etc. to see if OED might execute it.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+    mocha.describe('Token Validation', () => {
+        // Test through an endpoint that requires authentication
+        // Most endpoints use authMiddleware, so we'll test through a protected route
+        const PROTECTED_ENDPOINT = '/api/users'; // Requires admin auth
```
Just a general note that OED prefers comments on the line above.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,215 @@
+            expect(res).to.have.status(401);
+        });
+
+        mocha.it('should handle extremely long tokens (DoS prevention)', async () => {
```
I'm unclear how this differs from the hugeToken test above?

huss replied:
I see one is a bearer.

huss commented on src/server/test/routes/ciksParamsTest.js:
```diff
@@ -0,0 +1,85 @@
+
+    mocha.describe('Error Handling', () => {
+        mocha.it('should handle database connection errors gracefully', async () => {
+            const res = await chai.request(app)
```
Can you help me understand how a DB error would happen so this actually tests that?

omarraf replied:
The test as written doesn't actually simulate a real database error. It's just making a normal request and expecting it to work. To properly test database error handling, I'd need to either mock the database connection to throw an error or temporarily break the connection somehow. I'll remove this test since it's not actually testing error handling as intended.

huss replied:
Maybe change the description strings so they more accurately reflect the test?

huss commented on src/server/test/routes/compareReadingsParamsTest.js:
```diff
@@ -0,0 +1,321 @@
+            });
+
+            mocha.it('should reject invalid meter ID patterns', async () => {
+                const invalidPatterns = [
```
Does this apply across several routes so should be centralized? Also, could the map pattern be the base for this pattern with the added ones?

huss replied:
As per another comment, how about putting a TODO in the code for these ideas and leaving it for later?

huss commented on src/server/test/routes/compareReadingsParamsTest.js:
```diff
@@ -0,0 +1,321 @@
+            graphicUnitId: '1'
+        };
+
+        mocha.describe('URL Parameter Validation - meter_ids', () => {
```
I have not carefully checked but a number of these seems to be the type of tests that could apply across many routes with proper generalization. This was done for string, number, etc. Would it make sense to do that here?

huss reviewed (changes requested):
Thanks to @omarraf for another contribution. First, I would like to verify you are the only one who worked on this. I only see commits from you so just double checking. Second, I think the work is good. I read through some of the files to see how it is generally set up. I've made some comments in these files to give my thoughts. Given this is a draft, I wanted to get my thoughts back sooner rather than later and that some overall changes may be made, I did not do a full review. I hope what I did is useful and please let me know if there are specific files you would like me to look at that I did not. If anything is unclear, you have thoughts or need help then please just let me know.

omarraf commented:
Hi @huss, I really appreciate you getting back so quickly, I'm going to address all of the comments soon.  Just to answer your question real quick:

> I would like to verify you are the only one who worked on this

Yes, I ended up working on this alone, as the rest of my team was pulled onto other priorities.

omarraf commented:
Thanks again for taking the time to give your feedback @huss. I'll work through improving the PR and let you know if I need any help or need a review.

omarraf committed:
use standardized auth function

omarraf committed:
fix comment placement

omarraf committed:
Merge branch 'development' into systematic-testing

omarraf committed:
Merge branch 'development' into systematic-testing

omarraf committed:
create util file for constants

omarraf committed:
replace hardcoded maxLength values with constants for token and password validation for authenticator.js

omarraf committed:
remove redundant error handling tests for database connection in ciksParamsTest.js

omarraf committed:
add INTERNAL_SERVER_ERROR status code to HTTP_CODE constants

omarraf committed:
refactor: replace hardcoded maxLength values with constants in validation schemas and tests

omarraf committed:
refactor: update validation schemas to use constants for maxLength and minLength across multiple routes

omarraf committed:
refactor: add USERNAME_MIN_LENGTH, USERNAME_MAX_LENGTH, and PASSWORD_MIN_LENGTH to validation constants

omarraf committed:
add ISO date and duration validation functions to fix failing test

omarraf committed:
fix: add error handling for invalid timeInterval in readings routes

omarraf committed:
update login parameter validation to use constants for minLength and maxLength

omarraf committed:
add validateNoExtraFields function to enhance parameter validation in meters routes

omarraf committed:
fix: correct validation for group_id parameter in group 3D readings route

omarraf committed:
fix: add validation for timeInterval in readings count and raw meter routes

omarraf committed:
fix: update expected status codes in meters parameter validation tests

omarraf committed:
fix: update testInvalidField and validateNoExtraFields to support multiple expected status codes

omarraf committed:
fix: update status code expectations in authentication validation tests to include 429

omarraf committed:
update string length validations in conversions and meters parameter tests to use defined constants

omarraf committed:
update group creation validation tests to use defined constants for string length and improve extra fields validation

omarraf committed:
fix: update expected status codes in group and meter validation tests to include 429 for rate limiting

omarraf committed:
fix: skip parameter validation tests for curr_start, curr_end, shift, and oversized parameters until proper handling is implemented

omarraf committed:
fix: update username validation tests to handle undefined and length edge cases

omarraf commented:
Hi @huss, I've addressed the comments you provided, let me know if I'm missing anything or need me to clarify.

Changes:
- Consolidated route schemas to use shared validation constants
- Updated parameter tests to reuse the helper utilities, handle rate-limited responses
- Temporarily skipped compareReadings DoS checks (TODO added)

Some web/integration tests are failing in CI, which I didn't address in this PR, didn't do enough digging to figure out why

huss commented on src/server/test/routes/compareReadingsParamsTest.js:
```diff
@@ -0,0 +1,318 @@
+                expect(res.status).to.equal(400);
+            });
+
+            mocha.it.skip('should validate curr_start parameter', async () => { // TODO: re-enable once compareReadings rejects invalid ISO values without hitting DB
```
OED prefers comments on the line above the code. Applies in several places.

huss commented:
First, I want to apologize to @omarraf for the long delay in getting to this review. I made a mistake in my task tracking system and lost track of this PR. I'm very sorry this happened.

Second, there are 27 failing tests where most are for Obvius. I think many (all?) are due to changes in route checking. Would it be okay if I try to address them and push out changes?

omarraf commented:
No problem at all @huss. Feel free to make any changes

huss committed:
update string length consts & return code

- Make them alphabetical in declaration.
- Rename string so it comes first so all go together.
- Add more HTTP_CODE const responses and use in some places.
- Some formatting.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,250 @@
+            // Response should not reveal whether user exists or not
+            expect([400, 401]).to.include(res1.status);
+        });
+    });
```
I'm unclear how this tests rate limiting. Could you help me understand? Also, it raised another questing in my mind if we should rate limit in app.js for login attempts with a lower limit. What do you think?

omarraf replied:
Yeah looking back at it I think the comment is a bit misleading, it doesn't actually test rate limiting but just checks that all 10 requests fail with 400/401. I can either update the comment or add assertions for 429. Also I think adding a rate limit for logins is a good idea. We could add a seperate rate limiter instance in app.js just for /api/login with a lower threshold. I'm happy to add it in this PR or in a seperate issue.

huss replied:
I just opened issue #1564. Please leave a comment on that issue if you want to work on it so I can assign it to you.

Note this is now resolved but I'm not formally resolving so people can see this comment for a while.

huss commented on src/server/test/routes/authenticatorParamsTest.js:
```diff
@@ -0,0 +1,250 @@
+            // Response should not reveal whether user exists or not
+            expect([400, 401]).to.include(res1.status);
+        });
+    });
```
There seems a lot of similarities to the user test above. Do you think it would be possible to create a more general test to handle both?

omarraf replied:
There definitely is overlap, both test username/password validation, malicious inputs, and field length limits. I kept them separate mainly because of different context and test organization.  Authenticator tests validate middleware behavior (login endpoint), while user tests validate CRUD operations (/create, /edit, /delete)

That said, we could extract the shared test data (malicious username patterns, SQL injection strings, etc.) into a shared fixture file in src/server/test/util/. I can refactor it that way if need be

huss replied:
You have done a lot of work so it is not necessary for now. Other changes/consolidation of tests may become clear in the future. So, it will be put off for now but maybe a TODO comment in the code to indicate it might be desirable in the future would be nice.

huss commented on src/server/routes/compareReadings.js:
```diff
@@ -10,6 +10,16 @@ const moment = require('moment');
+const ISO_DURATION_REGEX = /^P(?!$)(\d+Y)?(\d+M)?(\d+D)?(T(\d+H)?(\d+M)?(\d+(\.\d+)?S)?)?$/;
+
+function isValidIsoDateTime(value) {
+	return typeof value === 'string' && moment.parseZone(value, moment.ISO_8601, true).isValid();
```
I see the value of checking if it is a string. However, doesn't the param validation do this before this runs?

omarraf replied:
Yeah I think you're right, I just added it to be safe but I can remove

huss replied:
I move we remove it since it is extra code to maintain.

huss commented on src/server/routes/compareReadings.js:
```diff
@@ -10,6 +10,16 @@ const moment = require('moment');
+}
+
+function isValidIsoDuration(value) {
+	return typeof value === 'string' && ISO_DURATION_REGEX.test(value);
```
Similar to above but why using ISO function to check.

huss committed:
minor changes to commit

- Use limits in test so always stay in sync.
- formatting

huss committed:
tab indenting & formatting of tests

huss commented on src/server/test/routes/readingsParamsTest.js:
```diff
@@ -0,0 +1,252 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+const { chai, mocha, expect, app, testUsers } = require('../common');
```
testUser not used.

huss commented on src/server/test/routes/readingsParamsTest.js:
```diff
@@ -0,0 +1,252 @@
+            expect([400, 500]).to.include(res.status);
+        });
+    });
+});
```
I think these endpoints should be used in the .get and .describe as possible.

huss committed:
use HTTP_CODE in tests

Did not change comments.
This does not fix other places such as routes.
Also, possible some lines might be too long now.

huss committed:
remove unused imports

huss committed:
convert more tests to use const on limits

Also formatting.

huss commented on src/server/routes/compareReadings.js:
```diff
@@ -106,10 +122,19 @@ function createRouter() {
+		const currEndRaw = req.query.curr_end;
+		const shiftRaw = req.query.shift;
+
+		if (!isValidIsoDateTime(currStartRaw) || !isValidIsoDateTime(currEndRaw) || !isValidIsoDuration(shiftRaw)) {
```
I don't have a problem with the new checks but I want to record that we don't do them in other open routes. This could be made into an issue for the future if you agree.

omarraf replied:
Yeah sounds good, I'd be open to working on the issue as well if need be

huss committed:
fix obvius route/test

The new param checking was stricter so the historical use of email
instead of the newer username was causing issues. This uses
username no matter which is used in the request.

huss committed:
add defaultAreaUnit to allowed params for preferences route

huss committed:
fix password length on user test

Note other values should be tested but another PR already
deals with that.

huss commented:
I have seen periodic failure of tests on GitHub and the latest [tests](https://github.com/OpenEnergyDashboard/OED/actions/runs/19619364268/job/56176825611?pr=1528) have 13 failures that are slightly different. The last 7 are 429/too many requests and I suspect the first 6 are OED's rate limiting error (the message is truncated so not certain). The first 6 might be because the new tests runs so fast and there are so many that it is hitting the limit for the first time. The last 7 are less clear but I did see  that src/server/routes/obvius.js has a new limit on multer but I have not checked that possibility out. Note these tests do not fail locally for me. This is going to have to be tracked down and I'm open to ideas.

huss commented on src/server/test/routes/ciksParamsTest.js:
```diff
@@ -0,0 +1,71 @@
+			const res = await chai.request(app)
+				.get('/api/ciks');
+
+			expect([HTTP_CODE.OK, HTTP_CODE.INTERNAL_SERVER_ERROR]).to.include(res.status);
```
I'm wondering why this and the next test would expect an error code.

omarraf replied:
I think I added this by mistake, it was probably because I was running the tests locally very frequently so some failed with server errors. I'll clean this up

huss reviewed (changes requested):
Thanks to @omarraf for the updated code. There are a lot of new tests and I appreciate the work in doing this. Overall, this is working well. I've tried to clear any resolved comments and added some more. I have not checked every test vs the route to verify all is being checked. I think that is a future task either after any changes are done or after this PR. One area with several comments is if tests could be generalized and the basic code reused across tests. For example, notes and maybe many strings are similar across params and routes. I don't know if you want to go into this now and it would be fine if it is done after this PR (not necessarily by you). Another point is that the code from the pending PR that started this process such as src/server/test/util/validationHelpers.js is not showing the git record of their having done the work. Did you change any of it or can it be included via that PR?

Please let me know your thoughts on the big picture items so we can decide how they will be addressed. There are other ones (such as using the HTTP codes throughout the coded as now done with route testing) that I plan to post as issues in the future and not in any comments on this PR.

Finally, I welcome your comments on the commits I made. I changed more than I might normally do because you have put in so much work and my review took so long. I hope this was okay.

omarraf commented:
Thanks for the thorough review and commits you made @huss , they're very helpful and I appreciate your time to improve the code. The changes you made look good to me and I'll address the rest of your requested changes soon.  Some of my thoughts on your questions:

> One area with several comments is if tests could be generalized and the basic code reused across tests

Definitely think it makes sense to consolidate similar validation patterns that could share test logic, but I think it'd be a great follow-up task rather than doing it in this PR since it would be a significant refactor and a seperate PR may be easier to review.

> Did you change any of it or can it be included via that PR?

Yeah, the validationHelpers.js I added directly from the previous PR on units.js when I first started this issue. I had left a comment at the top of the file conveying the original author and how the current file would be replaced by that PR. Since then I have built on their work and added several new helper functions (validateCommaSeparatedIdPatterns, expectValidCommaSeparatedIds, validateNoExtraFields, and modified testInvalidField to support multiple expected status codes). I want them to get credit for building the foundation so we can potentially take their original commit and add my modifications on top, or however you'd prefer to handle it.

For the bigger picture items I think we should focus on addressing the rate-limiting problem in CI and doing a thorough audit of all tests vs routes making sure we have complete coverage. For future work we can open up a new issue to do the test generalization and http codes throughout the codebase, but I'm happy to implement in this PR if need be.

omarraf committed:
Merge remote-tracking branch 'origin/development' into systematic-testing

Merge remote-tracking branch 'origin/development' into systematic-testing

omarraf committed:
update comments

omarraf committed:
Refactor CIKs parameter validation tests to assert only HTTP 200 status

omarraf committed:
Refactor readings parameter validation tests to use base endpoints for improved readability

omarraf commented:
@huss I addressed some of the comments you left, let me know if there's anything else I can change or clarify. 

Since this PR has been open for a while and covers a lot of ground, I'd like to wrap it up soon if there aren't any blocking issues. For the bigger refactoring items we discussed, I'm totally open to tackling those in follow-up issues. Same goes for the systematic http code audit and rate-limiting improvements in CI.

Let me know if there's anything else you'd like me to address in this PR. Happy to keep working on this effort either way.

omarraf commented:
Regarding the CI test failures, since most appear to be related to rate limiting, would it be possible to configure higher rate limits specifically for the test environment, or is there a preferred approach for handling this in CI? I wanted to check whether that would be considered good practice here.

huss commented on src/server/test/routes/readingsParamsTest.js:
```diff
@@ -0,0 +1,249 @@
+	const LINE_COUNT_BASE_ENDPOINT = '/api/readings/line/count/meters';
+	const RAW_READINGS_BASE_ENDPOINT = '/api/readings/line/raw/meter';
+
+	mocha.describe('GET /api/readings/line/count/meters/:meter_ids', () => {
```
To make sure the description stays in sync with the code, I think it might be better as `` `GET ${LINE_COUNT_BASE_ENDPOINT}/:meter_ids` ``. Same idea for the raw one below.

huss commented:
> For the bigger picture items I think we should focus on addressing the rate-limiting problem in CI

I tried playing with this for a little while but failed to get it to work. I tried to create a second app with a higher rate limit but did not get that to work, esp. given all the routes that would need to be redefined (this would not be a trivial solution). Maybe the limit on all tests could be changed. This isn't quite as good as only doing it for the desired ones but would be okay. It may be the case that middleware might fix this but did not look at that much. I'm open to ideas/solutions. If you want me to work on this more then let me know. I think it is the only remaining large issue that needs to be resolved so this can move forward.

huss reviewed (changes requested):
Thanks to @omarraf for the updated code. I'm sorry it took a little bit of time to get to this at the end of the year. Overall, this is looking very good. I made a few new comments and added to a few previous comments. My hope is those are easy to finalize. The issue of rate limiting causing failure is not yet resolved. I commented on my modest attempt so please see that comment to move forward. As always, let me know if I can help in any way.

omarraf committed:
Use template literals for endpoint URLs

omarraf committed:
remove unnecessary checks

omarraf committed:
add TODO comments

omarraf committed:
Add TODO and validation for request body in refresh route

omarraf committed:
adjust obvius param tests accept 429 responses

omarraf commented:
> I think it is the only remaining large issue that needs to be resolved so this can move forward.

I noticed that even after adjusting the Obvius tests to accept 429 responses, the rate limiting issue keeps appearing elsewhere. A temporary fix would be for tests to automatically look for and accept 429 as a valid response. As for other solutions since increasing the rate limit didn't work could we potentially disable it in the test environment? For middleware in app.js we could apply rate limiting when not testing. Not sure if this would be the desired outcome but just some thoughts. Also I addressed the comments you provided if I missed anything let me know, thanks!

omarraf commented:
Hey @huss just checking in on what some next steps might be. I can look more into the rate limiting issue and propose some ideas we can do, or did you have something else in mind? I was also thinking after these changes are fully integrated I can add to the [design docs](https://github.com/OpenEnergyDashboard/DesignDocs/tree/main/routeParamTesting) about changes introduced in this PR. I also took a look at the new issues you posted above but holding off on it for now until we finalize this.

huss commented:
> Hey @huss just checking in on what some next steps might be. I can look more into the rate limiting issue and propose some ideas we can do, or did you have something else in mind? I was also thinking after these changes are fully integrated I can add to the [design docs](https://github.com/OpenEnergyDashboard/DesignDocs/tree/main/routeParamTesting) about changes introduced in this PR. I also took a look at the new issues you posted above but holding off on it for now until we finalize this.

@omarraf Thanks for the comment:

- I don't have any direct ideas on the rate limiting issue and am open to ideas. Hopefully it will only change the rate limit for these or all tests so it should not have an impact on OED in regular environments. The rate limits are for security reasons outside of testing. It is important that a site could not easily have the rate limits changed/disabled in a production setting.
- It is always great when the design docs are updated for the current status. Often OED adds an update section rather than change the original so people can see the history. It is okay to update the original text if that is not important but an update is usually preferred.
- I'm hoping that once the rate limiting is fixed then the PR changes will not be too difficult for the rest.

Please let me know if I can help in any way or if you want to discuss rate limiting.

omarraf commented:
Hey @huss, for the rate limiting issue, I'm thinking of disabling it in the test environment by checking `NODE_ENV` in the middleware (app.js). Something like:

```javascript
if (process.env.NODE_ENV !== 'test') {
  app.use(rateLimiter);
}
```

This way rate limiting is only disabled when running tests, and can't be easily disabled in production. Curious on your thoughts on this approach.

huss commented:
@omarraf If you can safely detect when tests are running and switch off the rate limiting then that seems fine. A quick thought: Would it be easy to use a different rate limiter during testing that had larger limits? That would mean the app is still using one testing as is normally done during but it should not impact the tests. I think it is unlikely that turning it off would impact the app but this would mostly rule that out. However, it is okay either way. Thanks for moving on this.

omarraf committed:
testing seperate rate limit for test environment

omarraf committed:
Merge branch 'development' into systematic-testing

omarraf commented:
> different rate limiter during testing that had larger limits

Yeah this is a good idea, I went ahead and implemented a solution and looks like CI is passing. Let me know if this looks good to you or if there's anything you want me to change.

huss committed:
issue, comments, formatting

- The check if it was in the test environment to
modify the rate limit had an issue. The part that tested
for the OED_DB_TEST_DATABASE should not be there
since it is always set (even in production). I found the
rate limit was 100x even in production until I
remove that part of the test.
- I added some comments on the rate limiting
to make it very clear what is going on.
- Removed final comma in list per OED linting.

huss commented:
@omarraf Thank you for the updated code. I'm sorry it again took some time to get to this. Overall, the changes look good and all the comments within code are now resolved. Here is what I think are the final comments:

- src\server\test\routes\obviusParamsTest.js is checking for rate limits and skipping the usual test if that is returned. I see you added file limits but I'm assuming that is not the rate limit issue here. Given the changes to the rate limit in app.js for testing and the fact that I did not find any other rate limits on the obvius route, I'm trying to figure out if these are still needed. Could I get your thoughts?
- PR #1512 started this effort and you have expanded it greatly with this PR. I tried to see what in that PR is not in this PR and here is my list:
  - routes/unit.js: These changes are not in your PR and differs a little from how you changed routes.
  - test/routes/unitAddParamsTest.js: These tests are not in your PR and differ a little in how they are done. 
  - test/routes/unitEditParamsTest.js: These tests are not in your PR and differ a little in how they are done.
  - test/util/validationHelper.js: Similar to your changes but differ a little. 
  - test/util/routeTesting.js: The added parameter is renamed and moved in your PR.

My thinking is that if PR #1512 is merged then it may well create merge conflicts with your PR, cause some issues due to different function definitions and make unit param testing different than your newer work. Given this, I was thinking it might be best to manually add this work and incorporate your new methods. What do you think? Would you want to do that or should someone else?

Also, I pushed one commit to address an issue and some minor changes. I welcome any thoughts.

I think this is very close and I hope to get it wrapped up soon. Please let me know if I can help in any way.

omarraf commented:
> Could I get your thoughts?

It isn't needed. I added it to see if it might solve our problem but should've removed/not commit it.

I'm also willing to work on adding the units routes to this PR. I'll work on it shortly and also add the original author as a co-author on my commit. 

> I pushed one commit to address an issue and some minor changes. I welcome any thoughts.

Much appreciated!

omarraf committed:
remove rate limit checks obviusParamsTest

omarraf committed:
Add tests for unit route

Co-authored-by: maigbao <maigbao@users.noreply.github.com>

omarraf commented:
Hey @huss, added the units work to this PR. Let me know if there's anything else you want me to address

huss commented:
@omarraf Review and testing found that this is fine. As you noted, this is a large effort that has been ongoing so it is time to merge it.

Here are the items I have that are future work:

-  Test generalization and using more functions for testing.
- http codes throughout the codebase.
- Unit tests not checking max length.
- Systematically verify doing all the right checks for every route.

I wanted to touch base to see if you planned to work on any of these items. I was thinking that opening an issue for each one of these would be a good idea. Do you agree? I can do it unless you would like to.

Finally, @RamenAlmond is in the process of checking over this PR. This will continue even though I am about to merge this PR. Any ideas/issues found will be rolled into future issues/work.

huss reviewed (approved):
I want to thank @omarraf for this substantial work to improve our route testing/validation. I appreciate that you worked over time to address the comments and made the final solution very good. I have reviewed and tested the code and found it is fine. This is ready to merge. Congratulations on another accepted contribution to OED.

omarraf commented:
Thanks @huss. Glad we could get this merged.

I'm not sure if I'll work on those specific issues as of right now, I'll leave them for others to claim if they want to tackle them for the time being. If they're still open I may come back later and work on them. However I would still like to open them as issues if thats fine. Also, I was wondering on your thoughts if I create a PR updating/adding to the design documentation on route param testing regarding the changes in this PR?

huss commented:
@omarraf 

> I'm not sure if I'll work on those specific issues as of right now, I'll leave them for others to claim if they want to tackle them for the time being. If they're still open I may come back later and work on them.

You are welcome to work on any of these that remain or something else in the future.

> However I would still like to open them as issues if thats fine.

Feel free to open the issues as features/enhancements. Let me know if I can help in any way.

> Also, I was wondering on your thoughts if I create a PR updating/adding to the design documentation on route param testing regarding the changes in this PR?

You can fork/clone the [DesignDocs](https://github.com/OpenEnergyDashboard/DesignDocs) repo, make your edits and then open a PR to that repo. It is similar to what you did for this and will be reviewed similarly. Let me know if you have any questions.
